reference/src/layout/unions.md

```diff
@@ -38,3 +28,44 @@ contents are.
 Even if the offsets happen to be all 0, there might still be differences in the
 function call ABI.  If you need to pass unions by-value across an FFI boundary,
 you have to use `#[repr(C)]`.
+
+### Layout of `repr(C)` unions
+
+The layout of `repr(C)` unions follows the C layout scheme. Per sections
+[6.5.8.5] and [6.7.2.1.16] of the C11 specification, this means that the offset
+of every field is 0. Unsafe code can cast a pointer to the union to a field type
+to obtain a pointer to any field, and vice versa. 
+
+[6.5.8.5]: http://port70.net/~nsz/c/c11/n1570.html#6.5.8p5
+[6.7.2.1.16]: http://port70.net/~nsz/c/c11/n1570.html#6.7.2.1p16
+
+Since all fields are at offset 0, this implies that `repr(C)` unions do not have
+padding before or in-between their fields. They can, however, have trailing
+padding (see next example).
+
+Union fields of zero-size participate in the layout computation of the union. For example:
```

Comment:
This is just a consequence of prior statements, not a new normative clause, right? It should be marked as that.

Reply:
I don't think we have stated anything about repr(C) union fields of zero size. C doesn't have ZSTs, and in the section about struct layout we do specifically call out that when computing the layout of a `repr(C)` struct, if a field has zero-size, it participates in the layout computation in a specific way.

Reply:
There's a 3-line note there. That seems appropriate.

Reply:
So the note in the struct layout applies to union layout as well ? That feels to implicit to me, but I can live with that.

Reply:
No sorry that's not what I meant.

What I meant is, we can have a 3-line note here as well. But it seems entirely overkill to spend a section on this.

Reply:
Ok, that sounds good.

Reply:
@RalfJung the example shows trailing padding, so it cannot be removed. I can remove this one line, and remove the one line after the example, replacing them with a note similar to struct layout, but... then this section ends up being longer, which I'm not sure is what you meant.

Reply:
> the example shows trailing padding, so it cannot be removed

Wait, are you saying that ZST with alignment are the only way trailing padding can happen in a union? Or why is trailing padding not demonstrated with an example focused on trailing padding (the current example mixes both trailing padding and ZST involvement, it is not a good idea to demonstrate two things in one example)?

Reply:
Also it would help to have a structure like https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/structs-and-tuples.md#c-compatible-layout-repr-c, with "paragraph-level" sub headers.

Reply:
> Wait, are you saying that ZST with alignment are the only way trailing padding can happen in a union?

Nope, i'm just saying that the example currently demonstrates it with ZST, and the example is the bulk of this section. Removing / rephrasing the ZSTs bits, does not really make the section any shorter, which seemed to be the concern being raised:

> But it seems entirely overkill to spend a section on this.

Reply:
"this" = ZST.

I did not even realize that the example is doing both of these things. So, as a first step, can you split the section into two parts: `**Padding.**` and `**Zero-sized fields**.`? The first should then have an example for trailing padding does not use a zero-sized field.

Reply:
I'll try doing that.

Reply:
I've split the section into two subsections, one talking about padding, and the other one showing that zero-sized fields do not take up any size, but do still participate in the layout computation of the union.